### PR TITLE
[WIP] Channel safely

### PR DESCRIPTION
**Pull request by cppcooper**
Looking to rewrite via:

<https://github.com/DFHack/dfhack/pull/1879#discussion_r846696723> - @myk002 
and the one following.

### CODE REVIEW

**Comment by myk002** [comment]
Is this marked draft since the code is still in flux?

**Comment by cppcooper** [comment]
> Is this marked draft since the code is still in flux?

Effectively, yes. And because I have no idea when I'll get to this. I do think the tick event will be a nice way to assure safety during the job though.

**Comment by cppcooper** [comment]
So, to kick things off.. I'm thinking the new implementation will be aiming for this stuff. Based on previous discussion there may be another feature or two to add toggles for. The one that comes to mind now is whether to use simple processing, or to maybe do path finding.. or check where the dwarf is standing.. maybe also something like cancel-job if it becomes unsafe for a dwarf (or maybe animal, or visitor).

```
Multi-level channel projects can be dangerous, and managing the safety of your
dwarves throughout the completion of such projects can be difficult and time
consuming. This plugin keeps your dwarves safe (channeling) so you don't have to.

Usage
-----

``enable channel-safely``
    Start managing channel designations by marking/unmarking designations that
    are currently dangerous or will create a dangerous situation in the future.
    Designations under designations are the prime example of a dangerous
    situation that is mitigated.

``channel-safely require-vision <state>``
    Enables/disables requiring vision of a space to make decisions related to it.
    - state: 1/0 to enable/disable respectively
    When enabled the safety checks will only look below a designation if the
    dwarves already know what's below it. Therefore if you only want to disable
    designations after a dwarf has been sacrificed into the depths, this is the
    feature you want to enable.

``channel-safely allow-cheating <state>``
    Allows/disallows the plugin to cheat in order to safely channel designations
    as necessary.
    - state: 1/0 to enable/disable respectively
    Some designations may be permanently unsafe, or impossible to reach, and
    this feature can allow the instantaneous digging of such designations without
    a worker being involved in the task.

``channel-safely ignore <priority-threshold>``
    Configures the plugin to ignore designations equal to or above a specified priority.
```

**Comment by myk002** [comment]
could you unmark this as WIP when it's ready for review?

**Comment by cppcooper** [comment]
> could you unmark this as WIP when it's ready for review?

Sure, it could probably do with some code review now.

**Comment by Moth-Tolias** [comment]
`run once` as a single command feels weird. `--run-once` would be my preference

**Comment by myk002** [comment]
> `run once` as a single command feels weird. `--run-once` would be my preference

agreed, or just `run` or just `once`

**Comment by cppcooper** [comment]
nobody said `runonce` but I'll either do that or `run` when I get a chance to write a special case in the arg parser

File: plugins/channel-safely/channel-safely-plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+DFhackCExport command_result plugin_onupdate(color_ostream &out, state_change_event event) {
+    CSP::OnUpdate(out);
+    return DFHack::CR_OK;
+}
+
+command_result channel_safely(color_ostream &out, std::vector<std::string> &parameters) {
+    if (!parameters.empty()) {
+        if (parameters.size() >= 2 && parameters.size() <= 3) {
+            if (parameters[0] == "runonce") {
+                CSP::UnpauseEvent();
```
Comment (myk002) [inline plugins/channel-safely/channel-safely-plugin.cpp:588]: be sure to suspend core before reading/modifying game memory

**Review by myk002 — CHANGES_REQUESTED** [review]
I did not review all the logic (it would take splitting this PR into 5-10 PRs to cover that), but I left high-level comments.

File: docs/plugins/channel-safely.rst [hunk 2]
```diff
@@ -0,0 +1,63 @@
+channel-safely
+==============
+
+.. dfhack-tool::
+    :summary: Auto-manage channel designations to keep dwarves safe
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply docs/plugins/channel-safely.rst:5]:
```suggestion
    :summary: Auto-manage channel designations to keep dwarves safe.
```

File: docs/plugins/channel-safely.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    :tags: fort auto
+
+Multi-level channel projects can be dangerous, and managing the safety of your
+dwarves throughout the completion of such projects can be difficult and time
+consuming. This plugin keeps your dwarves safe (while channeling) so you don't
+have to. Now you can focus on designing your dwarven cities with the deep chasms
+they were meant to have.
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply docs/plugins/channel-safely.rst:12]:
```suggestion
consuming. This plugin keeps your dwarves safe (at least while channeling) so you don't
have to. Now you can focus on designing your dwarven cities with the deep chasms
they were meant to have.
```

File: docs/plugins/channel-safely.rst [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    or the tile below becomes open space the job will be paused or canceled (respectively).
+
+``channel-safely set ignore-threshold 3``
+    Configures the plugin to ignore designations equal to or above priority 3 designations.
+
+Features
+--------
+
+:require-vision:    Toggle whether the dwarves need vision of a tile before channeling to it can be deemed unsafe. (default: enabled)
+:monitor:           Toggle whether to monitor the conditions of active digs. (default: disabled)
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply docs/plugins/channel-safely.rst:52]:
```suggestion
``channel-safely enable monitor``
    Enables monitoring active channel digging jobs. Meaning that if another unit it present
    or the tile below becomes open space the job will be paused or canceled (respectively).

``channel-safely set ignore-threshold 3``
    Configures the plugin to ignore designations equal to or above priority 3 designations.

Features
--------

:require-vision:    Toggle whether the dwarves need vision of a tile before channeling to it can be deemed unsafe. (default: enabled)
:monitor:           Toggle whether to monitor the conditions of active digs. (default: disabled)
```

File: plugins/channel-safely/include/channel-manager.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    void destroy_groups() { groups.clear(); debug(); }
+    void manage_groups();
+    void manage_group(const df::coord &map_pos, bool set_marker_mode = false, bool marker_mode = false);
+    void manage_group(const Group &group, bool set_marker_mode = false, bool marker_mode = false);
+    bool manage_one(const Group &group, const df::coord &map_pos, bool set_marker_mode = false, bool marker_mode = false);
+    void mark_done(const df::coord &map_pos);
+    bool exists(const df::coord &map_pos) const { return groups.count(map_pos); }
+    void debug() {
+        DEBUG(groups).print(" DEBUGGING GROUPS:\n");
+        if (config.debug) {
```
Comment (myk002) [inline plugins/channel-safely/include/channel-manager.h:36]: do we need an additional state variable to track this or can we just use the existing `debugfilter`-managed categories?

File: docs/plugins/channel-safely.rst [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                    Runs on the refresh cycles. (default: disabled)
+
+Settings
+--------
+
+:refresh-freq:      The rate at which full refreshes are performed.
+                    This can be expensive if you're undertaking many mega projects. (default:600, twice a day)
+:monitor-freq:      The rate at which active jobs are monitored. (default:1)
+:ignore-threshold:  Sets the priority threshold below which designations are processed. You can set to 1 or 0 to
+                    effectively disable the scanning. (default: 5)
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply docs/plugins/channel-safely.rst:65]: "below which designations are processed"

File: plugins/channel-safely/include/inlines.h [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+            default:
+                designation.bits.dig = df::tile_dig_designation::No;
+                break;
+        }
+        Job::removeJob(job);
+    }
+}
+
+template<class Ctr1, class Ctr2, class Ctr3>
+void set_difference(const Ctr1 &c1, const Ctr2 &c2, Ctr3 &c3) {
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/include/inlines.h:173]: It is, I didn't like the args so I made a simpler implementation

File: plugins/channel-safely/channel-groups.cpp [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    // and so we begin iterating the neighbours
+    for (auto &neighbour: neighbors) {
+        // go to the next neighbour if this one doesn't have a group
+        if (!groups_map.count(neighbour)) {
+            TRACE(groups).print(" -> neighbour is not designated\n");
+            continue;
+        }
+        // get the group, since at least one exists... then merge any additional into that one
+        if (!group){
+            TRACE(groups).print(" -> group* has no valid state yet\n");
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-groups.cpp:36]: I see no problem

File: plugins/channel-safely/channel-safely-plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,709 @@
...
+#include <unordered_map>
+#include <unordered_set>
+
+// Debugging
+namespace DFHack {
+    DBG_DECLARE(channelsafely, plugin, DebugCategory::LINFO);
+    DBG_DECLARE(channelsafely, monitor, DebugCategory::LERROR);
+    DBG_DECLARE(channelsafely, manager, DebugCategory::LERROR);
+    DBG_DECLARE(channelsafely, groups, DebugCategory::LERROR);
+    DBG_DECLARE(channelsafely, jobs, DebugCategory::LERROR);
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp:79]: I use WARN() for the colour mostly

File: plugins/channel-safely/channel-safely-plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,709 @@
...
+                state = true;
+            } else if (parameters[0] == "disable") {
+                state = false;
+            } else if (parameters[0] == "set") {
+                set = true;
+            } else {
+                return DFHack::CR_WRONG_USAGE;
+            }
+            try {
+                if (parameters[1] == "debug") {
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp:603]: I don't know what that is, but is it? Notice that this doesn't set everything the same

File: plugins/channel-safely/channel-safely-plugin.cpp [hunk 11]
```diff
@@ -0,0 +1,709 @@
...
+                        // if this is a fresh start
+                        if (state && !config.monitor_active) {
+                            // we need a fresh start
+                            CSP::active_workers.clear();
+                        }
+                    }
+                } else if (parameters[1] == "refresh-freq" && set && parameters.size() == 3) {
+                    config.refresh_freq = std::abs(std::stol(parameters[2]));
+                } else if (parameters[1] == "monitor-freq" && set && parameters.size() == 3) {
+                    config.monitor_freq = std::abs(std::stol(parameters[2]));
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp:675]: I may not agree with the choice to use 0, but I will defend the player's right to make bad decisions. It effectively would just mean any tick related code is running as frequently as it possibly can.

File: plugins/channel-safely/channel-manager.cpp [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+#include <inlines.h>
+
+#include <modules/EventManager.h> //hash function for df::coord
+#include <df/block_square_event_designation_priorityst.h>
+
+/**
+blocks[48][96][135]: <map_block: 0x7fff8e587200>
+blocks[48][96][135].default_liquid.hidden:  false
+blocks[48][96][135].designation[10][0].hidden:  false
+ * */
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-manager.cpp:12]: /shrug
from very long ago. I may have wanted a note of the designation.hidden layout.. maybe the liquid as well. Dunno

[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-manager.cpp:12]:
```suggestion

```

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 1 above)
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp:588]: oh? why is this? I haven't had any troubles so far.. I think

File: plugins/channel-safely/include/channel-manager.h (hunk 5 above)
Comment (cppcooper) [reply plugins/channel-safely/include/channel-manager.h:36]: `config.debug` it is necessary for that exact use iirc (enables the code if any debugging is going on)

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/channel-safely/channel-safely-plugin.cpp:603]: https://docs.dfhack.org/en/stable/docs/tools/debug.html?highlight=debug%20filter%20#debug

File: plugins/channel-safely/channel-groups.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/channel-safely/channel-groups.cpp:36]: I think it would only be a problem if you needed to get trace info from a Windows user who was hitting a bug. I don't think NDEBUG is ever defined when building on Linux

File: docs/plugins/channel-safely.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (myk002) [reply docs/plugins/channel-safely.rst:65]: Oi ok human parsing error disregard

File: plugins/channel-safely/channel-groups.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-groups.cpp:36]: sounds like a problem for another PR

File: plugins/channel-safely/include/inlines.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/include/inlines.h:173]: in fact, I'm gonna make it even simpler when I remove the inner loop whenever I boot into linux.

File: plugins/channel-safely/include/tile-cache.h [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+#include <modules/Maps.h>
+#include <df/coord.h>
+#include <df/tiletype.h>
+
+#include <map>
+
+class TileCache {
+private:
+    TileCache() = default;
+    std::map<df::coord, df::tiletype> locations;
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/include/tile-cache.h:12]:
```suggestion
    std::unordered_map<df::coord, df::tiletype> locations;
```

File: plugins/channel-safely/include/tile-cache.h [hunk 14]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <modules/Maps.h>
+#include <df/coord.h>
+#include <df/tiletype.h>
+
+#include <map>
```
Comment (cppcooper) [inline plugins/channel-safely/include/tile-cache.h:7]:
```suggestion
#include <unordered_map>
```

File: plugins/channel-safely/include/tile-cache.h [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        locations.erase(pos);
+    }
+
+    bool hasChanged(const df::coord &pos, const df::tiletype &type) {
+        if (locations.count(pos)) {
+            if (type != locations.find(pos)->second){
+                return true;
+            }
+        }
+        return false;
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/include/tile-cache.h:33]: you changed the comparison
```suggestion
        return locations.count(pos) && type != locations[pos].second;
```

File: plugins/channel-safely/channel-safely-plugin.cpp [hunk 16]
```diff
@@ -0,0 +1,709 @@
...
+        !Lua::PushModulePublic(out, L, "plugins.dig-now", "dig_now_tile"))
+        return false;
+
+    Lua::Push(L, map_pos);
+
+    if (!Lua::SafeCall(out, L, 1, 1))
+        return false;
+
+    return lua_toboolean(L, -1);
+
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp:154]: This indentation is wild, and I can't tell what's going on. I'll add the suggestion with the batch, but I'm gonna re-indent this when I boot into my regular environment

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 1 above)
Comment (myk002) [reply plugins/channel-safely/channel-safely-plugin.cpp:584]: it can create race conditions, and can mess up the lua stack if you ever call into Lua from a thread that doesn't suspend the core

File: plugins/channel-safely/include/channel-manager.h (hunk 5 above)
Comment (myk002) [reply plugins/channel-safely/include/channel-manager.h:36]: that's what I'm confused about -- doesn't `debugfilter` already track the verbosity level for all your categories? why do you need another tracker here?

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/channel-safely/channel-safely-plugin.cpp:599]: and `debugfilter` allows you to set categories independently.

File: plugins/channel-safely/channel-groups.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/channel-safely/channel-groups.cpp:36]: agreed, I didn't mean to imply that changes to Debug.h belong in this PR.

**Comment by myk002** [comment]
You need a hash template param for the unordered maps. see https://github.com/DFHack/dfhack/blob/bf6d569ea2f813587400033fe21f24ecea2a3a96/plugins/blueprint.cpp#L183 for an example

**Comment by cppcooper** [comment]
> You need a hash template param for the unordered maps. see
> 
> https://github.com/DFHack/dfhack/blob/bf6d569ea2f813587400033fe21f24ecea2a3a96/plugins/blueprint.cpp#L183
> for an example

nah I just need to include the eventmanager's hash functions

File: plugins/channel-safely/include/channel-manager.h (hunk 5 above)
Comment (cppcooper) [reply plugins/channel-safely/include/channel-manager.h]: It was because I wanted it to get to this function if any debugging was taking place

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 1 above)
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp]: So this is only important for lua related code?

File: plugins/channel-safely/channel-safely-plugin.cpp [hunk 17]
```diff
@@ -0,0 +1,709 @@
...
+};
+
+enum SettingConfigData {
+    REFRESH_RATE,
+    MONITOR_RATE,
+    IGNORE_THRESH,
+    FALL_THRESH
+};
+
+// dig-now.cpp
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp:111]: It is infrequent, but I'd rather it be in the Maps module (or another module)

File: plugins/channel-safely/channel-groups.cpp [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+    /* We need to add map_pos to an existing group if possible...
+     * So what we do is we look at neighbours to see if they belong to one or more existing groups
+     * If there is more than one group, we'll be merging them
+     */
+    df::coord neighbors[8];
+    get_neighbours(map_pos, neighbors);
+    Group* group = nullptr;
+    int group_index = -1;
+
+    DEBUG(groups).print("    add(" COORD ")\n", COORDARGS(map_pos));
```
Comment (myk002) [inline plugins/channel-safely/channel-groups.cpp:37]: nice

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/channel-safely/channel-safely-plugin.cpp:111]: SGTM -- do you want to move it in this PR or a separate one?

File: plugins/channel-safely/channel-safely-plugin.cpp [hunk 19]
```diff
@@ -0,0 +1,666 @@
...
+
+df::coord simulate_area_fall(const df::coord &pos) {
+    df::coord neighbours[8]{};
+    get_neighbours(pos, neighbours);
+    df::coord lowest = simulate_fall(pos);
+    for (auto p : neighbours) {
+        if (p.z < lowest.z) {
+            lowest = p;
+        }
+    }
```
Comment (myk002) [inline plugins/channel-safely/channel-safely-plugin.cpp:134]: if you aren't simulating falling for the neighbors, when will this for loop ever change the value for lowest?

**Review by myk002 — APPROVED** [review]
Overall LGTM. Is it ready for merge?

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 19 above)
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp:134]: oops, brain must have been faster than my fingers

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (cppcooper) [reply plugins/channel-safely/channel-safely-plugin.cpp:111]: separate

File: plugins/channel-safely/channel-safely-plugin.cpp (hunk 1 above)
Comment (myk002) [reply plugins/channel-safely/channel-safely-plugin.cpp]: AIUI, it's important for all memory-accessing code. It just shows visible errors for Lua interactions. Everything else just corrupts silently.

**Review by myk002 — APPROVED** [review]

File: docs/plugins/channel-safely.rst [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+dwarves throughout the completion of such projects can be difficult and time
+consuming. This plugin keeps your dwarves safe (at least while channeling) so you don't
+have to. Now you can focus on designing your dwarven cities with the deep chasms
+they were meant to have.
+
+Usage
+-----
+
+::
+    enable channel-safely
```
Comment (myk002) [inline docs/plugins/channel-safely.rst:18]: needs blank line here. current text has parsing errors:
image

File: docs/plugins/channel-safely.rst [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+Settings
+--------
+
+:refresh-freq:      The rate at which full refreshes are performed.
+                    This can be expensive if you're undertaking many mega projects. (default:600, twice a day)
+:monitor-freq:      The rate at which active jobs are monitored. (default:1)
+:ignore-threshold:  Sets the priority threshold below which designations are processed. You can set to 1 or 0 to
+                    effectively disable the scanning. (default: 5)
+:fall-threshold:    Sets the fall threshold beyond which is considered unsafe. (default: 1)
```
[the remark this answers is not in the document]
Comment (cppcooper) [reply docs/plugins/channel-safely.rst:74]: They all are saved yes. I thought it was mentioned, I'll update the doc

File: docs/plugins/channel-safely.rst [hunk 22]
```diff
@@ -27,6 +27,8 @@ and designations deemed unsafe will be put into :wiki:`Marker Mode <Designations
 Each time a channel designation is completed its group status is checked, and if the group
 is complete pending groups below are made active again.
 
+Settings will persist according to the save game being used.
```
Comment (myk002) [inline docs/plugins/channel-safely.rst:30]: I find this wording a bit hard to follow. Autolabor says this: `Once you have enabled it in a fortress, it stays enabled until you explicitly disable it, even if you save and reload your game.` which is more verbose, but also more clear.

Comment (cppcooper) [reply docs/plugins/channel-safely.rst:30]: I find that wording misleading. The plugin doesn't stay enabled, just the features.

Comment (cppcooper) [reply docs/plugins/channel-safely.rst:30]: I could:
```
Features and settings once set will persist until you change them, even if you save and reload your game.
```
Does that work for you?

Comment (myk002) [reply docs/plugins/channel-safely.rst:30]: that's good. and plugin enablement will be persisted soon (tm) once I get that working for all plugins.

Comment (cppcooper) [reply docs/plugins/channel-safely.rst:30]: does it have an off switch?

Comment (cppcooper) [reply docs/plugins/channel-safely.rst:30]:
```suggestion
Features and settings once set will persist until you change them, even if you save and reload your game.
```

**Review by myk002 — APPROVED** [review]

File: plugins/channel-safely/channel-groups.cpp [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+    std::set<df::coord> last_jobs;
+    for (auto &pos : jobs) {
+        last_jobs.emplace(pos);
+    }
+    jobs.load_channel_jobs();
+    // transpose channel jobs to
+    std::set<df::coord> new_jobs;
+    std::set<df::coord> gone_jobs;
+    set_difference(last_jobs, jobs, gone_jobs);
+    set_difference(jobs, last_jobs, new_jobs);
```
[the remark this answers is not in the document]
Comment (ab9rf) [reply plugins/channel-safely/channel-groups.cpp:133]: it's fine, i just didn't see it in `inlines.h`
